### PR TITLE
Changing the experiments API to session API

### PR DESCRIPTION
**Pull request by balodja**
Please, guys, take a look.

### CODE REVIEW

**Comment by hombit** [comment]
I converted to draft, because I was about to merge it accidentally

**Comment by balodja** [comment]
@matwey , @hombit , is the realization now better?

File: coniferest/session/callback.py [hunk 1]
```diff
@@ -50,3 +54,72 @@ def __call__(self, metadata, data, session) -> None:
...
+            return
+
+        ordering = np.argsort(session.scores)
+        index, = np.where(ordering == session.current)
+        self._trace.append(ordering[:index[0] + self._anomalies + 1])
+
+
+def make_autonomous_experiment(
+        metadata: 'np.ndarray[Label]',
+        budget: int = None) -> AutonomousExperiment:
```
Comment (hombit) [inline coniferest/session/callback.py:120]: type should be `Optional[int]`

File: coniferest/session/callback.py [hunk 2]
```diff
@@ -50,3 +54,72 @@ def __call__(self, metadata, data, session) -> None:
...
+
+        if not self._save_trace:
+            return
+
+        ordering = np.argsort(session.scores)
+        index, = np.where(ordering == session.current)
+        self._trace.append(ordering[:index[0] + self._anomalies + 1])
+
+
+def make_autonomous_experiment(
```
Comment (hombit) [inline coniferest/session/callback.py:118]: Please add a docstring

File: coniferest/session/callback.py [hunk 3]
```diff
@@ -50,3 +54,72 @@ def __call__(self, metadata, data, session) -> None:
...
         if self.iteration >= self.budget:
             session.terminate()
+
+
+def autonomous_decider_callback(metadata: Label, data: np.ndarray, session: 'Session') -> Label:
+    "Use metadata as already defined labels for making decision."
+    return metadata
+
+
+class AutonomousExperiment:
```
Comment (hombit) [inline coniferest/session/callback.py:64]: The purpose of this class is not clear from the name and the docstring. I don't understand the name, but I cannot think about anything better. @matwey, do you have better ideas?
More documentation will be useful, but could go to another PR.

**Review by hombit — APPROVED** [review]
Thank you, it has much more sense for me! I approve it with few minor comments. I'd like to see @matwey's review after @balodja make this PR to be "Ready for review" (not draft)

**Comment by hombit** [comment]
Probably, `AutonomousExperiment` could be decomposed into smaller callbacks: `TerminateAfter(budget)` we [already have](https://github.com/snad-space/coniferest/blob/6674bf0fce497b78bf21d4ce774eddd6e2334741/coniferest/session/callback.py#L40), `terminate_after_all_detected` we [have in tests](https://github.com/snad-space/coniferest/blob/6674bf0fce497b78bf21d4ce774eddd6e2334741/tests/test_session.py#L70) and a new callback for tracing.

**Comment by hombit** [comment]
`trajectory` sounds like a useful thing for `Session`, maybe we need to implement it there? We currently have trajectory implicitly, at least for the case of no prior labels, it is just `list(session.known_labels)`

File: coniferest/session/callback.py (hunk 1 above)
Comment (balodja) [reply coniferest/session/callback.py:120]: You are right. Maybe `int | None`?

**Comment by balodja** [comment]
> Probably, `AutonomousExperiment` could be decomposed into smaller callbacks: `TerminateAfter(budget)` we [already have](https://github.com/snad-space/coniferest/blob/6674bf0fce497b78bf21d4ce774eddd6e2334741/coniferest/session/callback.py#L40), `terminate_after_all_detected` we [have in tests](https://github.com/snad-space/coniferest/blob/6674bf0fce497b78bf21d4ce774eddd6e2334741/tests/test_session.py#L70) and a new callback for tracing.

I thought about modifying the `TerminateAfter` callback to include the termination decision on number of anomalies found. But I was too afraid you wouldn't like it.

File: coniferest/session/callback.py (hunk 1 above)
Comment (hombit) [reply coniferest/session/callback.py:120]: It was introduced in Python 3.10 https://peps.python.org/pep-0604/

**Comment by hombit** [comment]
> I thought about modifying the `TerminateAfter` callback to include the termination decision on number of anomalies found. But I was too afraid you wouldn't like it.

We can have these two callbacks separately and combine them when creating a session. However, I don't think it is a bad idea to add a limit of anomalies to `TerminateAfter`, but it just looks like not very useful thing for real experts. I'd even say that real expert would prefer to limit number of non-anomalies instead

File: coniferest/session/callback.py [hunk 4]
```diff
@@ -50,3 +54,72 @@ def __call__(self, metadata, data, session) -> None:
         self.iteration += 1
         if self.iteration >= self.budget:
             session.terminate()
+
+
+def autonomous_decider_callback(metadata: Label, data: np.ndarray, session: 'Session') -> Label:
```
[the remark this answers is not in the document]
Comment (balodja) [reply coniferest/session/callback.py:59]: And what should I use then? `lambda: label, data, session: label`? Better have a name for that.

File: coniferest/session/callback.py (hunk 2 above)
Comment (balodja) [reply coniferest/session/callback.py:119]: Done

File: coniferest/session/callback.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hombit) [reply coniferest/session/callback.py:59]: I think lambda is a bit longer but much more verbose. I'm ok with this function as well

File: src/coniferest/session/oracle.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            # Session.metadata is labels, so we just use this candidate metadata is the true label
+            decision_callback=lambda label, _features, _self: label,
+            on_decision_callbacks=on_decision_callbacks,
+        )
+
+    @property
+    def trace(self) -> List[np.ndarray[int]]:
+        if self._tracer is None:
+            raise AttributeError("Construct with tracer=True to use .trace")
+        return self._tracer.trace
```
Comment (matwey) [inline src/coniferest/session/oracle.py:69]: Since trace is a kind of public attribute, can we document what trace is?

**Comment by matwey** [comment]
Also, can you rebase this PR into single commit to avoid "Remove XXX" changes reverting previous commit?

File: src/coniferest/session/oracle.py (hunk 5 above)
Comment (hombit) [reply src/coniferest/session/oracle.py]: I removed Tracer
